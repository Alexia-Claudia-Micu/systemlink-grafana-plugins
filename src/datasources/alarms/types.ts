import { DataQuery, DataSourceJsonData } from '@grafana/schema'
import { QueryBuilderField } from 'smart-webcomponents-react';

export interface AlarmsQuery extends DataQuery {
  queryKind: AlarmQueryType;
}

export enum AlarmQueryType {
  ListAlarms = "List Alarms",
  AlarmCount = "Alarm Count",
  AlarmTrend = "Alarm Trend",
}

export interface QBField extends QueryBuilderField {
  lookup?: {
    readonly?: boolean;
    dataSource: Array<{
      label: string,
      value: string
    }>;
  },
}

export interface AlarmDataSourceOptions extends DataSourceJsonData {
}
