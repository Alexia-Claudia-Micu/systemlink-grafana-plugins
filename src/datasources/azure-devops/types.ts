import { DataQuery } from '@grafana/schema'

export interface AzureDevopsQuery extends DataQuery {
  type: string;
};
