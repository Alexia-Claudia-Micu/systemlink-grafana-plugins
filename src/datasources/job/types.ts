import { DataQuery } from '@grafana/schema'

export enum QueryType {
  Summmary = "SUMMARY",
  Whatever = "WHATEVER"
}

export interface JobQuery extends DataQuery {
  type: QueryType
}

export interface JobSummary {
  activeCount: number;
  failedCount: number;
  succeededCount: number;
}