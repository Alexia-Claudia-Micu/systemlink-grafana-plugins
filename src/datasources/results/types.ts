import { DataQuery } from '@grafana/schema';

export interface ResultsQuery extends DataQuery {
  outputType: OutputType;
  properties?: ResultsProperties[];
  orderBy?: string;
  descending?: boolean;
  useTimeRange?: boolean;
  useTimeRangeFor?: string;
  recordCount?: number;
}

export enum OutputType {
  Data = 'Data',
  TotalCount = 'Total Count',
}

export enum UseTimeRange {
  Started = 'Started',
  Updated = 'Updated',
}

export enum ResultsProperties {
    id = 'ID',
    programName = 'PROGRAM_NAME',
    serialNumber = 'SERIAL_NUMBER',
    systemId = 'SYSTEM_ID',
    status = 'STATUS',
    totalTimeInSeconds = 'TOTAL_TIME_IN_SECONDS',
    startedAt= 'STARTED_AT',
    updatedAt = 'UPDATED_AT',
    partNumber = 'PART_NUMBER',
    dataTableIds= 'DATA_TABLE_IDS',
    fileIds = 'FILE_IDS',
    hostName = 'HOST_NAME',
    operator = 'OPERATOR',
    keywords = 'KEYWORDS',
    properties = 'PROPERTIES',
    statusTypeSummary = 'STATUS_TYPE_SUMMARY',
    workspace = 'WORKSPACE',
}

export const OrderBy = [
  {
    value: 'ID',
    label: 'ID',
    description: `ID of the result`,
  },
  {
    value: 'STARTED_AT',
    label: 'Started At',
    description: `Started At of the result`,
  },
  {
    value: 'UPDATED_AT',
    label: 'Updated At',
    description: `Updated At of the result`,
  },
  {
    value: 'PROGRAM_NAME',
    label: 'Program Name',
    description: `Program Name of the result`,
  },
  {
    value: 'SYSTEM_ID',
    label: 'System ID',
    description: `System ID of the result`,
  },
  {
    value: 'HOST_NAME',
    label: 'Host Name',
    description: `Host Name of the result`,
  },
  {
    value: 'OPERATOR',
    label: 'Operator',
    description: `Operator of the result`,
  },
  {
    value: 'SERIAL_NUMBER',
    label: 'Serial Number',
    description: `Serial Number of the result`,
  },
   {
    value: 'PART_NUMBER',
    label: 'Part Number',
    description: `Part Number of the result`,
  },
  {
    value: 'PROPERTIES',
    label: 'Properties',
    description: `Properties of the result`,
  },
  {
    value: 'TOTAL_TIME_IN_SECONDS ',
    label: 'Total Time In Seconds',
    description: `Total Time In Seconds of the result`,
  }
];

export const ResultsPropertiesOptions = {
    ID: 'id',
    PROGRAM_NAME: 'programName',
    SERIAL_NUMBER: 'serialNumber',
    SYSTEM_ID: 'systemId',
    STATUS: 'status',
    TOTAL_TIME_IN_SECONDS: 'totalTimeInSeconds',
    STARTED_AT: 'startedAt',
    UPDATED_AT: 'updatedAt',
    PART_NUMBER: 'partNumber',
    DATA_TABLE_IDS: 'dataTableIds',
    FILE_IDS: 'fileIds',
    HOST_NAME: 'hostName',
    OPERATOR: 'operator',
    KEYWORDS: 'keywords',
    PROPERTIES: 'properties',
    STATUS_TYPE_SUMMARY: 'statusTypeSummary',
    WORKSPACE: 'workspace',
}

export interface StatusHttp {
  statusType: string;
  statusName: string;
}

export interface StatusTypeSummaryHttp {
  [status: string]: number;
}

export interface ResultsResponseProperties {
  status?: StatusHttp;
  startedAt?: string;
  updatedAt?: string;
  programName: string;
  id: string;
  systemId: string;
  hostName: string;
  operator: string;
  partNumber: string;
  serialNumber: string;
  totalTimeInSeconds: number;
  keywords: string[];
  properties: { [key: string]: string };
  fileIds: string[];
  statusTypeSummary?: StatusTypeSummaryHttp;
  workspace: string;
  dataTableIds: string[];
}

export interface QueryResultsResponse {
  results: ResultsResponseProperties[];
  continuationToken: string;
  totalCount: number;
}
