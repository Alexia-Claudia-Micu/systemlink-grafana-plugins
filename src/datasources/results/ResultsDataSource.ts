import { DataFrameDTO, DataQueryRequest, DataSourceInstanceSettings, FieldType, MetricFindValue, TestDataSourceResponse } from '@grafana/data';
import { BackendSrv, TemplateSrv, getBackendSrv, getTemplateSrv } from '@grafana/runtime';
import { DataSourceBase } from 'core/DataSourceBase';
import { OutputType, QueryResultsHttpResponse, QueryStepsHttpResponse, ResultsQuery as TestResultsQuery, ResultsQueryType, ResultsVariableQuery, ResultsProperties, QueryDataTablesHttpResponse, DataTablesProperties, ResultsVariableQueryType } from './types';

export class TestResultsDataSource extends DataSourceBase<TestResultsQuery> {
  constructor(
    readonly instanceSettings: DataSourceInstanceSettings,
    readonly backendSrv: BackendSrv = getBackendSrv(),
    readonly templateSrv: TemplateSrv = getTemplateSrv()
  ) {
    super(instanceSettings, backendSrv, templateSrv);
  }

  baseUrl = this.instanceSettings.url;
  queryResultsUrl = this.baseUrl + '/nitestmonitor/v2/query-results';
  queryTablesUrl = this.baseUrl + '/nidataframe/v1/tables';
  getResultsCountUrl = this.baseUrl + '/nitestmonitor/v2/results';
  queryStepsUrl = this.baseUrl + '/nitestmonitor/v2/query-steps';

  private timeRange: { [key: string]: string } = {
    Started: 'startedAt',
    Updated: 'updatedAt',
  }
  private fromDateString = '${__from:date}';
  private toDateString = '${__to:date}';

  defaultQuery = {
    type: ResultsQueryType.Results,
    recordCount: 1000,
    outputType: OutputType.Data,
    useTimeRange: false,
  };

  async getWorkspaceNames(): Promise<{ label: string, value: string }[]> {
    const workspaces = await this.getWorkspaces();
    const workspaceOptions = workspaces.map(w => ({ label: w.name, value: w.id }));
    const variables = this.getVariableOptions().map(v => ({ label: v, value: v }))
    workspaceOptions?.unshift(...variables)
    return workspaceOptions;
  }

  async queryResults(filter: string, projection?: ResultsProperties[], orderBy?: string, recordCount = 1000, descending = false, returnCount = false): Promise<QueryResultsHttpResponse> {
    return await this.post<QueryResultsHttpResponse>(`${this.queryResultsUrl}`, {
      filter: filter,
      projection: projection ? projection : undefined,
      orderBy: orderBy,
      orderByDescending: descending,
      take: recordCount,
      returnCount: returnCount
    }).then(response => response);
  }

  async queryDataTables(id: string, orderBy?: string, workspace?: string[], recordCount = 1000, descending = false): Promise<QueryDataTablesHttpResponse> {
    return await this.get<QueryDataTablesHttpResponse>(`${this.queryTablesUrl}`, {
      id: id,
      orderBy: orderBy,
      descending: descending,
      take: recordCount,
      workspace: workspace
    }).then(response => response);
  }

  async queryTestResultValues(field: string, startsWith: string,): Promise<string[]> {
    if (!startsWith || startsWith.startsWith('$')) {
      return this.getVariableOptions();
    }
    const data = { field, startsWith };
    const values = await getBackendSrv().post(this.baseUrl + '/nitestmonitor/v2/query-result-values', data);
    return values.slice(0, 20).filter((value: string) => value);
  }

  private getVariableOptions = () => {
    return getTemplateSrv()
      .getVariables()
      .map(v => '$' + v.name);
  };

  async queryStepsValues(field: string, startsWith: string): Promise<string[]> {
    if (!startsWith || startsWith.startsWith('$')) {
      return this.getVariableOptions();
    }
    const data = { field, startsWith };
    const values = await getBackendSrv().post(this.baseUrl + '/nitestmonitor/v2/query-step-values', data);
    return values.slice(0, 20).filter((value: string) => value);
  }

  async runQuery(query: TestResultsQuery, options: DataQueryRequest): Promise<DataFrameDTO> {
    if (query.type === ResultsQueryType.Results) {
      let queryByFilter = query.queryBy ? query.queryBy : '';
      if (query.useTimeRange && query.useTimeRangeFor) {
        const timeFilter = `(${this.timeRange[query.useTimeRangeFor]} >= \"${this.fromDateString}\" && ${this.timeRange[query.useTimeRangeFor]} <= \"${this.toDateString}\")`
        queryByFilter = queryByFilter ? `${queryByFilter} && ${timeFilter}` : timeFilter
      }
      const variableReplacedFilter = getTemplateSrv().replace(queryByFilter, options.scopedVars)
      if (query.type === ResultsQueryType.Results) {
        const responseData = (await this.queryResults(variableReplacedFilter, query.properties?.length ? query.properties : undefined, query.orderBy, query.recordCount, query.descending, true));
        if (query.outputType === OutputType.Data) {
          const response = responseData.results;

          const filteredFields = query.properties?.filter((field: ResultsProperties) => Object.keys(response[0]).includes(field)) || [];

          const status = response.map(result => result['status']?.statusType);

          const fields = filteredFields.map((field) => {
            const fieldType = field === ResultsProperties.updated ? FieldType.time : FieldType.string;
            const values = response.map(m => m[field]);
            if (field === ResultsProperties.status.toLowerCase()) {
              return { name: field, values: status.flat(), type: FieldType.string };
            }
            return { name: field, values, type: fieldType };
          });

          return {
            refId: query.refId,
            fields: fields,
          };

        }
        else {
          return {
            refId: query.refId,
            fields: [
              { name: 'Total count', values: [responseData.totalCount] }
            ],
          };
        }
      }
      else {
        const responseData = (await this.queryResults(variableReplacedFilter, [ResultsProperties.dataTablesIds], query.orderBy, query.recordCount, query.descending, true));
        const dataTableIds = responseData.results.map(r => r.dataTableIds).flat();
        const dataTables = await this.queryDataTables(dataTableIds.join(','), query.orderBy, query.workspace ? [query.workspace] : undefined, query.recordCount, query.descending); console.log(query.properties)
        if (query.outputType === OutputType.Data) {
          if (dataTables.dataTables.length === 0) {
            return {
              refId: query.refId,
              fields: [
                { name: 'ID', values: [dataTableIds] }
              ]
            };
          } else {
            const fields = (query.properties as unknown as DataTablesProperties[])?.map((field) => {
              const values = dataTables!.dataTables.map(m => m[field]);
              return { name: field, values };
            })
            return {
              refId: query.refId,
              fields: fields

            }
          }
        } else {
          return {
            refId: query.refId,
            fields: [
              { name: 'Total count', values: [dataTableIds.length] }
            ]
          }
        }
      }
    } else {
      let queryByStepFilter = query.stepFilter;
      if (query.useTimeRange && query.useTimeRangeFor) {
        const timeFilter = `(${this.timeRange[query.useTimeRangeFor]} >= \"${this.fromDateString}\" && ${this.timeRange[query.useTimeRangeFor]} <= \"${this.toDateString}\")`
        queryByStepFilter = queryByStepFilter ? `${queryByStepFilter} && ${timeFilter}` : timeFilter
      }

      const variableReplacedResultFilter = getTemplateSrv().replace(query.resultFilter, options.scopedVars);
      const variableReplacedStepFilter = getTemplateSrv().replace(queryByStepFilter, options.scopedVars);
      const response = await this.post<QueryStepsHttpResponse>(this.queryStepsUrl, {
        resultFilter: variableReplacedResultFilter,
        projection: query.properties ? query.properties : undefined,
        orderBy: query.orderBy,
        descending: query.descending,
        take: query.recordCount,
        filter: variableReplacedStepFilter,
      });
      if (!query.measurementAsEntries) {
        return {
          refId: query.refId,
          fields: [
            { name: 'Step Name', values: response.steps.map(step => step.name) },
            { name: 'Step Type', values: response.steps.map(step => step.stepType) },
            { name: 'Step Id', values: response.steps.map(step => step.stepId) },
            { name: 'Parent Id', values: response.steps.map(step => step.parentId) },
            { name: 'Result Id', values: response.steps.map(step => step.resultId) },
            { name: 'Path', values: response.steps.map(step => step.path) },
            { name: 'Path Ids', values: response.steps.map(step => step.pathIds) },
            { name: 'Status', values: response.steps.map(step => step.status) },
            { name: 'Total Time (s)', values: response.steps.map(step => step.totalTimeInSeconds) },
            { name: 'Started At', values: response.steps.map(step => step.startedAt) },
            { name: 'Updated At', values: response.steps.map(step => step.updatedAt) },
            { name: 'Inputs', values: response.steps.map(step => step.inputs) },
            { name: 'Outputs', values: response.steps.map(step => step.outputs) },
            { name: 'Data Model', values: response.steps.map(step => step.dataModel) },
            { name: 'Data', values: response.steps.map(step => step.data) },
            { name: 'Has Children', values: response.steps.map(step => step.hasChildren) },
            { name: 'Workspace', values: response.steps.map(step => step.workspace) },
            { name: 'Keywords', values: response.steps.map(step => step.keywords) },
            { name: 'Properties', values: response.steps.map(step => step.properties) },
          ],
        };
      } else {
        const measurementName = response.steps.map(step => step.data.parameters.map(param => param.name))
        const measurementStatus = response.steps.map(step => step.data.parameters.map(param => param.status))
        const measurementValue = response.steps.map(step => step.data.parameters.map(param => param.measurement))
        return {
          refId: query.refId,
          fields: [
            { name: 'Step Name', values: response.steps.map(step => step.name) },
            { name: 'Step Type', values: response.steps.map(step => step.stepType) },
            { name: 'Step Id', values: response.steps.map(step => step.stepId) },
            { name: 'Parent Id', values: response.steps.map(step => step.parentId) },
            { name: 'Result Id', values: response.steps.map(step => step.resultId) },
            { name: 'Path', values: response.steps.map(step => step.path) },
            { name: 'Path Ids', values: response.steps.map(step => step.pathIds) },
            { name: 'Status', values: response.steps.map(step => step.status) },
            { name: 'Total Time (s)', values: response.steps.map(step => step.totalTimeInSeconds) },
            { name: 'Started At', values: response.steps.map(step => step.startedAt) },
            { name: 'Updated At', values: response.steps.map(step => step.updatedAt) },
            { name: 'Inputs', values: response.steps.map(step => step.inputs) },
            { name: 'Outputs', values: response.steps.map(step => step.outputs) },
            { name: 'Data Model', values: response.steps.map(step => step.dataModel) },
            { name: 'MeasurementName', values: measurementName.flat() },
            { name: 'MeasurementValue', values: measurementValue.flat() },
            { name: 'MeasurementStatus', values: measurementStatus.flat() },
            { name: 'Has Children', values: response.steps.map(step => step.hasChildren) },
            { name: 'Workspace', values: response.steps.map(step => step.workspace) },
            { name: 'Keywords', values: response.steps.map(step => step.keywords) },
            { name: 'Properties', values: response.steps.map(step => step.properties) },
          ],
        };
      }
    }
  }

  async metricFindQuery({ type, queryBy, workspace, resultFilter, stepFilter, properties }: ResultsVariableQuery, options: DataQueryRequest): Promise<MetricFindValue[]> {
    const workspaceFilter = `workspace = (\"${workspace}\")`;
    if (type === ResultsVariableQueryType.Results) {
      let queryByFilter = workspace && queryBy ? `${workspaceFilter} && ${queryBy}` : `${workspaceFilter}` || queryBy;
      const variableReplacedFilter = getTemplateSrv().replace(queryByFilter, options.scopedVars);
      if (type === ResultsVariableQueryType.Results) {
        const responseData = (await this.queryResults(variableReplacedFilter, [ResultsProperties.programName], undefined, 1000, false, false)).results;
        return responseData.map(frame => ({ text: frame.programName, value: frame.programName, }));
      } else {
        const responseData = (await this.queryResults(variableReplacedFilter, [ResultsProperties.dataTablesIds], undefined, 1000, false, false)).results;
        const dataTableIds = responseData.map(frame => frame.dataTableIds).flat();
        return dataTableIds.map(dataTableId => ({ text: dataTableId, value: dataTableId }));
      }
    } else {
      const variableReplacedResultFilter = getTemplateSrv().replace(resultFilter, options.scopedVars);
      const variableReplacedStepFilter = getTemplateSrv().replace(stepFilter, options.scopedVars);
      const response = await this.post<QueryStepsHttpResponse>(this.queryStepsUrl, {
        resultFilter: variableReplacedResultFilter,
        take: 1000,
        filter: variableReplacedStepFilter,
      });
      return response.steps.map(step => ({ text: step.name, value: step.name }));
    }
  }

  shouldRunQuery(query: TestResultsQuery): boolean {
    return true;
  }

  async testDatasource(): Promise<TestDataSourceResponse> {
    // TODO: Implement a health and authentication check
    await this.backendSrv.get(this.baseUrl + '/bar');
    return { status: 'success', message: 'Data source connected and authentication successful!' };
  }
}
