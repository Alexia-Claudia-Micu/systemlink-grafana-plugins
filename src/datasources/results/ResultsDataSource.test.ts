import { MockProxy } from 'jest-mock-extended';
import { ResultsDataSource } from './ResultsDataSource';
import { OutputType, QueryResultsResponse, ResultsProperties, ResultsPropertiesOptions, ResultsQuery, UseTimeRange } from './types';
import { BackendSrv, getTemplateSrv } from '@grafana/runtime';
import { createFetchError, createFetchResponse, getQueryBuilder, requestMatching, setupDataSource } from 'test/fixtures';
import { Field } from '@grafana/data';

const mockQueryResulltsResponse: QueryResultsResponse = {
  results: [
    {
      status: {
        statusType: 'PASSED',
        statusName: 'Passed',
      },
      startedAt: '2018-05-07T18:58:05.219692Z',
      updatedAt: '2023-10-23T22:30:03.833396Z',
      programName: 'My Program Name',
      id: '000007fb-aa87-4ab9-9757-6568e7893c33',
      systemId: '20FRS0MQ67--SN-R90N9L4N--MAC-54-AA-75-CE-B7-FE',
      hostName: 'My-Host',
      operator: 'admin',
      partNumber: 'cRIO-9050',
      serialNumber: '123-456',
      totalTimeInSeconds: 29.9,
      keywords: ['keyword1', 'keyword2'],
      properties: {
        Mass: '',
        StartTime: '10-19-2022',
        MaxVoltage: '4.02',
        MetrackID1: 'None',

      },
      statusTypeSummary: {
        "LOOPING": 0,
        "SKIPPED": 0,
        "CUSTOM": 0,
        "DONE": 0,
        "PASSED": 1,
        "FAILED": 0,
        "RUNNING": 0,
        "WAITING": 0,
        "TERMINATED": 0,
        "ERRORED": 0,
        "TIMED_OUT": 0
      },
      fileIds: ['5e30934193c608046851acb2'],
      dataTableIds: ['62333547f7521f2f2f4675e5'],
      workspace: '846e297h-a007-47ac-9fc2-fac07eab240e',
    },
  ],
  continuationToken: '',
  totalCount: 1
};

jest.mock('@grafana/runtime', () => ({
  getTemplateSrv: jest.fn()
}));

let datastore: ResultsDataSource, backendServer: MockProxy<BackendSrv>

describe('ResultsDataSource', () => {
  beforeEach(() => {
    (getTemplateSrv as jest.Mock).mockReturnValue({
        replace: jest.fn((value) => value.replace('${__from:date}', 'replacedDate'))
      });
    [datastore, backendServer] = setupDataSource(ResultsDataSource);

    backendServer.fetch
    .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-results', method: 'POST' }))
    .mockReturnValue(
      createFetchResponse<QueryResultsResponse>(mockQueryResulltsResponse)
    );
  })

  describe('testDataSource', () => {
    test('returns success', async () => {
      backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-results-test-api', method: 'GET' }))
      .mockReturnValue(createFetchResponse('testData'));

      const response = await datastore.testDatasource();

      expect(response.status).toEqual('success');
    });

    test('bubbles up exception', async () => {
      backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-results-test-api', method: 'GET' }))
      .mockReturnValue(createFetchError(400));

      await expect(datastore.testDatasource())
      .rejects
      .toThrow('Request to url "/nitestmonitor/v2/query-results-test-api" failed with status code: 400. Error message: "Error"');
    });
  });

  describe('queryResults', () => {
    test('returns data when there are valid queries', async () => {
      const response = await datastore.queryResults();

      expect(response).toMatchSnapshot();
    });

    test('raises an error returns API fails', async () => {
      backendServer.fetch
        .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-results' }))
        .mockReturnValue(createFetchError(400));

      await expect(datastore.queryResults())
        .rejects
        .toThrow('Request to url "/nitestmonitor/v2/query-results" failed with status code: 400. Error message: "Error"');
    });
  });

  describe('query', () => {
    test('returns data for valid data-output-type query', async () => {
      const query = buildQuery({
        refId: 'A',
        outputType: OutputType.Data,
        properties: [
          ResultsPropertiesOptions.ID,
          ResultsPropertiesOptions.PROGRAM_NAME,
          ResultsPropertiesOptions.SERIAL_NUMBER,
          ResultsPropertiesOptions.SYSTEM_ID,
          ResultsPropertiesOptions.STATUS_TYPE_SUMMARY
        ] as ResultsProperties[],
        orderBy: ResultsPropertiesOptions.ID,
        descending: false,
        useTimeRange: true,
        useTimeRangeFor: UseTimeRange.Updated,
        recordCount: 1
      });

      const response = await datastore.query(query);

      expect(response.data).toHaveLength(1);
      expect(response.data).toMatchSnapshot();
    });

    test('returns total count for valid total count output type queries', async () => {
      const query = buildQuery({
        refId: 'A',
        outputType: OutputType.TotalCount,
        useTimeRange: true,
        useTimeRangeFor: UseTimeRange.Updated
      });

      const response = await datastore.query(query);

      expect(response.data).toHaveLength(1);
      expect(response.data).toMatchSnapshot();
    });

    test('returns no data when QueryResults API returns empty array', async () => {
      backendServer.fetch
        .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-results', method: 'POST' }))
        .mockReturnValue(
          createFetchResponse(
            {
              results: [],
              continuationToken: null,
              totalCount: 0
            } as unknown as QueryResultsResponse
          )
        );

      const query = buildQuery();
      const response = await datastore.query(query);

      expect(response.data).toMatchSnapshot();
    });

    test('returns no data when Query Results returns error', async () => {
        backendServer.fetch
          .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-results' }))
          .mockReturnValue(createFetchError(400));

        const query = buildQuery(
          {
            refId: 'A',
            outputType: OutputType.Data,
            properties: [
              ResultsPropertiesOptions.ID,
              ResultsPropertiesOptions.PROGRAM_NAME,
              ResultsPropertiesOptions.SERIAL_NUMBER,
              ResultsPropertiesOptions.SYSTEM_ID,
            ] as ResultsProperties[],
            orderBy: ResultsPropertiesOptions.ID,
            descending: false,
            useTimeRange: true,
            useTimeRangeFor: UseTimeRange.Updated,
            recordCount: 1
          },
        );

        await expect(datastore.query(query))
        .rejects
        .toThrow('Request to url "/nitestmonitor/v2/query-results" failed with status code: 400. Error message: "Error"');
    });

    test('should convert properties to Grafana fields', async () => {
        const query = buildQuery(
          {
            refId: 'A',
            outputType: OutputType.Data,
            properties: [
              ResultsPropertiesOptions.ID,
              ResultsPropertiesOptions.PROGRAM_NAME,
              ResultsPropertiesOptions.SERIAL_NUMBER,
              ResultsPropertiesOptions.SYSTEM_ID,
              ResultsPropertiesOptions.PROPERTIES,
              ResultsPropertiesOptions.STATUS,
              ResultsPropertiesOptions.STATUS_TYPE_SUMMARY
            ] as ResultsProperties[],
            orderBy: ResultsPropertiesOptions.ID,
            descending: false,
            useTimeRange: true,
            useTimeRangeFor: UseTimeRange.Updated,
            recordCount: 1
          },
        );

        const response = await datastore.query(query);

        const fields = response.data[0].fields as Field[];
        expect(fields).toMatchSnapshot();
    });

    test('should handle null and undefined properties', async () => {
        backendServer.fetch
          .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-results' }))
          .mockReturnValue(createFetchResponse({
            results: [
              {
                id: '1',
                properties: null
              }
            ], continuationToken: null, totalCount: 1
          } as unknown as QueryResultsResponse));

        const query = buildQuery(
          {
            refId: 'A',
            outputType: OutputType.Data,
            properties: [
              ResultsPropertiesOptions.PROPERTIES
            ] as ResultsProperties[],
            orderBy: undefined
          },
        );

        const response = await datastore.query(query);
        const fields = response.data[0].fields as Field[];
        expect(fields).toEqual([
          { name: 'properties', values: [""], type: 'string' },
        ]);
    });
  });

  const buildQuery = getQueryBuilder<ResultsQuery>()({
    refId: 'A',
    properties: [
      ResultsPropertiesOptions.ID,
      ResultsPropertiesOptions.PROGRAM_NAME,
      ResultsPropertiesOptions.SERIAL_NUMBER,
      ResultsPropertiesOptions.SYSTEM_ID,
    ] as ResultsProperties[],
    orderBy: undefined
  });
});
