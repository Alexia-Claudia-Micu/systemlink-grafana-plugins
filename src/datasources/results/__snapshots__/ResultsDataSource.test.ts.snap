// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsDataSource query returns data for valid data output type queries 1`] = `
[
  {
    "fields": [
      {
        "name": "id",
        "type": "string",
        "values": [
          "000007fb-aa87-4ab9-9757-6568e7893c33",
        ],
      },
      {
        "name": "programName",
        "type": "string",
        "values": [
          "My Program Name",
        ],
      },
      {
        "name": "serialNumber",
        "type": "string",
        "values": [
          "123-456",
        ],
      },
      {
        "name": "systemId",
        "type": "string",
        "values": [
          "20FRS0MQ67--SN-R90N9L4N--MAC-54-AA-75-CE-B7-FE",
        ],
      },
      {
        "name": "statusTypeSummary",
        "type": "string",
        "values": [
          "{"LOOPING":0,"SKIPPED":0,"CUSTOM":0,"DONE":0,"PASSED":1,"FAILED":0,"RUNNING":0,"WAITING":0,"TERMINATED":0,"ERRORED":0,"TIMED_OUT":0}",
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`ResultsDataSource query returns data for valid data-output-type query 1`] = `
[
  {
    "fields": [
      {
        "name": "id",
        "type": "string",
        "values": [
          "000007fb-aa87-4ab9-9757-6568e7893c33",
        ],
      },
      {
        "name": "programName",
        "type": "string",
        "values": [
          "My Program Name",
        ],
      },
      {
        "name": "serialNumber",
        "type": "string",
        "values": [
          "123-456",
        ],
      },
      {
        "name": "systemId",
        "type": "string",
        "values": [
          "20FRS0MQ67--SN-R90N9L4N--MAC-54-AA-75-CE-B7-FE",
        ],
      },
      {
        "name": "statusTypeSummary",
        "type": "string",
        "values": [
          "{"LOOPING":0,"SKIPPED":0,"CUSTOM":0,"DONE":0,"PASSED":1,"FAILED":0,"RUNNING":0,"WAITING":0,"TERMINATED":0,"ERRORED":0,"TIMED_OUT":0}",
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`ResultsDataSource query returns no data when Query Products returns no data 1`] = `[]`;

exports[`ResultsDataSource query returns no data when QueryResults API returns empty array 1`] = `[]`;

exports[`ResultsDataSource query returns total count for valid total count output type queries 1`] = `
[
  {
    "fields": [
      {
        "name": "Total count",
        "values": [
          1,
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`ResultsDataSource query should convert properties to Grafana fields 1`] = `
[
  {
    "name": "id",
    "type": "string",
    "values": [
      "000007fb-aa87-4ab9-9757-6568e7893c33",
    ],
  },
  {
    "name": "programName",
    "type": "string",
    "values": [
      "My Program Name",
    ],
  },
  {
    "name": "serialNumber",
    "type": "string",
    "values": [
      "123-456",
    ],
  },
  {
    "name": "systemId",
    "type": "string",
    "values": [
      "20FRS0MQ67--SN-R90N9L4N--MAC-54-AA-75-CE-B7-FE",
    ],
  },
  {
    "name": "properties",
    "type": "string",
    "values": [
      "{"Mass":"","StartTime":"10-19-2022","MaxVoltage":"4.02","MetrackID1":"None"}",
    ],
  },
  {
    "name": "status",
    "type": "string",
    "values": [
      "PASSED",
    ],
  },
  {
    "name": "statusTypeSummary",
    "type": "string",
    "values": [
      "{"LOOPING":0,"SKIPPED":0,"CUSTOM":0,"DONE":0,"PASSED":1,"FAILED":0,"RUNNING":0,"WAITING":0,"TERMINATED":0,"ERRORED":0,"TIMED_OUT":0}",
    ],
  },
]
`;

exports[`ResultsDataSource queryResults returns data when there are valid queries 1`] = `
{
  "continuationToken": "",
  "results": [
    {
      "dataTableIds": [
        "62333547f7521f2f2f4675e5",
      ],
      "fileIds": [
        "5e30934193c608046851acb2",
      ],
      "hostName": "My-Host",
      "id": "000007fb-aa87-4ab9-9757-6568e7893c33",
      "keywords": [
        "keyword1",
        "keyword2",
      ],
      "operator": "admin",
      "partNumber": "cRIO-9050",
      "programName": "My Program Name",
      "properties": {
        "Mass": "",
        "MaxVoltage": "4.02",
        "MetrackID1": "None",
        "StartTime": "10-19-2022",
      },
      "serialNumber": "123-456",
      "startedAt": "2018-05-07T18:58:05.219692Z",
      "status": {
        "statusName": "Passed",
        "statusType": "PASSED",
      },
      "statusTypeSummary": {
        "CUSTOM": 0,
        "DONE": 0,
        "ERRORED": 0,
        "FAILED": 0,
        "LOOPING": 0,
        "PASSED": 1,
        "RUNNING": 0,
        "SKIPPED": 0,
        "TERMINATED": 0,
        "TIMED_OUT": 0,
        "WAITING": 0,
      },
      "systemId": "20FRS0MQ67--SN-R90N9L4N--MAC-54-AA-75-CE-B7-FE",
      "totalTimeInSeconds": 29.9,
      "updatedAt": "2023-10-23T22:30:03.833396Z",
      "workspace": "846e297h-a007-47ac-9fc2-fac07eab240e",
    },
  ],
  "totalCount": 1,
}
`;
